solr/core/src/java/org/apache/solr/util/circuitbreaker/GlobalCircuitBreakerManager.java

```diff
@@ -0,0 +1,156 @@
+import org.slf4j.LoggerFactory;
+
+public class GlobalCircuitBreakerManager implements ClusterPropertiesListener {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private static final ObjectMapper mapper = SolrJacksonAnnotationInspector.createObjectMapper();
+  private final CircuitBreakerRegistry cbRegistry = new CircuitBreakerRegistry();
+  private final GlobalCircuitBreakerFactory factory;
+  private final CoreContainer coreContainer;
+  private int currentConfigHash;
+
+  public CircuitBreakerRegistry getCircuitBreakerRegistry() {
+    return cbRegistry;
+  }
+
+  public GlobalCircuitBreakerManager(CoreContainer coreContainer) {
+    super();
+    this.factory = new GlobalCircuitBreakerFactory(coreContainer);
+    this.coreContainer = coreContainer;
+  }
+
+  private static class GlobalCircuitBreakerConfig {
+    static final String CIRCUIT_BREAKER_CLUSTER_PROPS_KEY = "circuit-breakers";
+    @JsonProperty Map<String, CircuitBreakerConfig> configs = new ConcurrentHashMap<>();
+
+    @JsonProperty
+    Map<String, Map<String, CircuitBreakerConfig>> hostOverrides = new ConcurrentHashMap<>();
+
+    static class CircuitBreakerConfig {
+      @JsonProperty Boolean enabled = false;
+      @JsonProperty Boolean debugMode = false;
+      @JsonProperty Double updateThreshold = Double.MAX_VALUE;
+      @JsonProperty Double queryThreshold = Double.MAX_VALUE;
+    }
+  }
+
+  // for registering global circuit breakers set in clusterprops
+  @Override
+  public boolean onChange(Map<String, Object> properties) {
+    try {
+      GlobalCircuitBreakerConfig nextConfig = processConfigChange(properties);
+      if (nextConfig != null) {
+        registerCircuitBreakers(nextConfig);
+      }
+    } catch (Exception e) {
+      if (log.isWarnEnabled()) {
+        // don't break when things are misconfigured
+        log.warn("error parsing global circuit breaker configuration {}", e);
+      }
+    }
+    return false;
+  }
+
+  private GlobalCircuitBreakerConfig processConfigChange(Map<String, Object> properties)
+      throws IOException {
+    Object cbConfig = properties.get(GlobalCircuitBreakerConfig.CIRCUIT_BREAKER_CLUSTER_PROPS_KEY);
+    GlobalCircuitBreakerConfig globalCBConfig = null;
+    if (cbConfig != null) {
+      byte[] configInput =
+          Utils.toJSON(
+              properties.get(GlobalCircuitBreakerConfig.CIRCUIT_BREAKER_CLUSTER_PROPS_KEY));
+
+      int nextHashCode = Arrays.hashCode(configInput);
+      if (this.currentConfigHash == nextHashCode) {
+        return null;
+      }
+      this.currentConfigHash = nextHashCode;
+
+      if (configInput != null && configInput.length > 0) {
+        globalCBConfig = mapper.readValue(configInput, GlobalCircuitBreakerConfig.class);
+      }
+    }
+    return globalCBConfig;
+  }
+
+  private void registerCircuitBreakers(GlobalCircuitBreakerConfig gbConfig) throws Exception {
+    this.cbRegistry.deregisterAll();
```

Comment:
Curious (relevant to, but not specific to, this PR): why call `deregisterAll()` instead of `close()`? Tbh I'm unclear why the `deregisterAll()` method exists (as distinct from `close()`, but given that it appears to currently only be called from test code, I'd be wary of introducing a call to it in application code.

Reply:
`deregisterAll()` just seemed to be clearer for the purpose I was using it for. `close()` to me seems a bit more final, while I'm still using the registry, I just want it cleared out. I have no qualms with changing this to `close()`, though!